src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -299,38 +309,137 @@ public int bytesLeft()
+    }
+
+    @Override
+    public long write(ByteBuffer[] srcs) throws IOException
+    {
+      return addToOffset(channel.write(srcs));
     }
 
-    public int addToOffset(int numBytesWritten)
+    public int addToOffset(long numBytesWritten)
     {
       if (numBytesWritten > bytesLeft()) {
         throw new ISE("Wrote more bytes[%,d] than available[%,d]. Don't do that.", numBytesWritten, bytesLeft());
       }
 
       currOffset += numBytesWritten;
 
-      return numBytesWritten;
+      return Ints.checkedCast(numBytesWritten);
+    }
+
+    @Override
+    public boolean isOpen()
+    {
+      return channel.isOpen();
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+      channel.close();
+    }
+  }
+
+  private static class BufferedWritableByteChannel
+      implements GatheringByteChannel
+  {
+    private static final int MAX_BUFFER_SIZE = 65536;
+    private static final int DEFAULT_BUFFER_SIZE = 8192;  // default of buffered output stream
+
+    private final WritableByteChannel channel;
+    private final ByteBuffer buffer;
+
+    BufferedWritableByteChannel(WritableByteChannel channel)
+    {
+      this(channel, DEFAULT_BUFFER_SIZE);
+    }
+
+    BufferedWritableByteChannel(WritableByteChannel channel, int buffer)
+    {
+      this.channel = channel;
+      this.buffer = ByteBuffer.allocateDirect(Math.min(buffer, MAX_BUFFER_SIZE));
+    }
+
+    public int write(ByteBuffer src) throws IOException
+    {
+      int position = src.position();
+      while (src.hasRemaining()) {
+        ByteBuffer toWrite = src;
+        int bytesToWrite = Math.min(buffer.remaining(), src.remaining());
+        if (bytesToWrite != src.remaining()) {
+          toWrite = src.duplicate();
+          toWrite.limit(toWrite.position() + bytesToWrite);
+        }
+        buffer.put(toWrite);
+        src.position(toWrite.position());
+        if (!buffer.hasRemaining()) {
+          flush(false);
+        }
+      }
+      return src.position() - position;
+    }
+
+    @Override
+    public long write(ByteBuffer[] srcs, int offset, int length) throws IOException
+    {
+      long written = 0;
+      for (int i = offset; i < length; i++) {
+        written += write(srcs[i]);
+      }
+      return written;
+    }
+
+    @Override
+    public long write(ByteBuffer[] srcs) throws IOException
+    {
+      return write(srcs, 0, srcs.length);
+    }
+
+    private void flush(boolean flushAll) throws IOException
+    {
+      buffer.flip();
+      do {
+        channel.write(buffer);
+      } while (flushAll && buffer.hasRemaining());
+
+      if (buffer.hasRemaining()) {
+        buffer.compact();
+      } else {
+        buffer.clear();
+      }
     }
 
     @Override
     public boolean isOpen()
     {
-      return open;
+      return channel.isOpen();
     }
 
     @Override
     public void close() throws IOException
     {
-      open = false;
-      out.close();
+      try {
```

Comment:
I suggest to use TWR "to close" idiom here as more concise and benefitting Throwable suppression rather than losing them:
```
try (Closeable toClose = channel) {
  if (buffer.position() != 0) {
    flush(true);
  }
}
```